File: app/src/main/java/com/jamieadkins/gwent/card/list/BaseCardListFragment.java
package com.jamieadkins.gwent.card.list;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;

import com.jamieadkins.gwent.R;
import com.jamieadkins.gwent.base.BaseFragment;
import com.jamieadkins.gwent.card.CardFilter;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * UI fragment that shows a list of the users decks.
 */

public abstract class BaseCardListFragment extends BaseFragment
        implements CardsContract.View {
    private CardsContract.Presenter mCardsPresenter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(getLayoutId(), container, false);
        setupViews(rootView);
        return rootView;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        setupFilterMenu(menu, inflater);
    }

    public int getLayoutId() {
        return R.layout.fragment_card_list;
    }

    @Override
    public void onStart() {
        super.onStart();
        onLoadData();
    }

    @Override
    public void onCardFilterUpdated() {
        getRecyclerViewAdapter().clear();
        onLoadData();
    }

    @Override
    public void onLoadData() {
        super.onLoadData();
        onLoadCardData();
    }

    public void onLoadCardData() {
        CardFilter cardFilter = getCardFilter();
        mCardsPresenter.getCards(cardFilter)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(getObserver());
    }

    @Override
    public void setPresenter(CardsContract.Presenter presenter) {
        mCardsPresenter = presenter;
    }
}

File: app/src/main/java/com/jamieadkins/gwent/card/list/BaseCardListFragment.kt
package com.jamieadkins.gwent.card.list

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import com.jamieadkins.commonutils.ui.RecyclerViewItem

import com.jamieadkins.gwent.R
import com.jamieadkins.gwent.base.BaseFragment
import com.jamieadkins.gwent.bus.RxBus
import com.jamieadkins.gwent.bus.SnackbarBundle
import com.jamieadkins.gwent.bus.SnackbarRequest
import com.jamieadkins.gwent.model.GwentCard
import android.icu.lang.UCharacter.GraphemeClusterBreak.T



/**
 * UI fragment that shows a list of the users decks.
 */

abstract class BaseCardListFragment<T : CardsContract.View> : BaseFragment<T>(), CardsContract.View {

    lateinit var emptyView: View

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(layoutId, container, false)
        emptyView = rootView.findViewById(R.id.empty)
        setupViews(rootView)
        return rootView
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        setupFilterMenu(menu, inflater)
    }

    override fun showCards(cards: MutableList<GwentCard>) {
        if (cards.size > 0) {
            context?.let { context ->
                val preferences = PreferenceManager.getDefaultSharedPreferences(context)
                val locale = preferences.getString(context.getString(R.string.pref_locale_key), context.getString(R.string.default_locale))
                val items = mutableListOf<RecyclerViewItem>()
                items.addAll(cards.sortedBy { it.name[locale] })
                showItems(items)

                emptyView.visibility = View.GONE
                mRecyclerView.visibility = View.VISIBLE

                // Have to post this since we have to wait for diff util to finish.
                mRecyclerView.post({
                    mRecyclerView.scrollToPosition(0)
                })
            }
        } else {
            emptyView.visibility = View.VISIBLE
            mRecyclerView.visibility = View.GONE
        }

    }

    override fun showEmptyView() {

    }

    open val layoutId: Int = R.layout.fragment_card_list
}
